File: src/file-view-tracker.ts
import * as vscode from 'vscode';
import { ImportTreeItem } from './tree/items';
import { getCurrentTreeProvider } from './import-manager';

// Keep track of currently visible files
let currentlyViewedFiles: Set<string> = new Set<string>();

/**
 * Initialize the file view tracker
 */
export function initializeFileViewTracker(context: vscode.ExtensionContext): void {
    // Listen for active editor changes to track which files are being viewed
    context.subscriptions.push(
        vscode.window.onDidChangeActiveTextEditor(editor => {
            updateCurrentlyViewedFiles(editor);
        })
    );

    // Initialize with currently open editor
    const activeEditor = vscode.window.activeTextEditor;
    if (activeEditor && activeEditor.document && activeEditor.document.uri && activeEditor.document.uri.fsPath) {
        currentlyViewedFiles.add(activeEditor.document.uri.fsPath);
    }

    // Also listen to document opening/closing to handle edge cases
    context.subscriptions.push(
        vscode.workspace.onDidOpenTextDocument(document => {
            // Note: We only add if this document becomes the active editor
            // This is handled by the onDidChangeActiveTextEditor event
        })
    );
}

/**
 * Update the tracking when the active editor changes
 */
function updateCurrentlyViewedFiles(editor: vscode.TextEditor | undefined): void {
    // Clear all currently viewed files
    currentlyViewedFiles.clear();

    // Add the currently active editor if it exists
    if (editor && editor.document && editor.document.uri && editor.document.uri.fsPath) {
        currentlyViewedFiles.add(editor.document.uri.fsPath);
    }

    // Update tree view to reflect current state
    updateTreeViewIndicators();
}

/**
 * Check if a file path is currently being viewed
 */
export function isFileCurrentlyViewed(filePath: string): boolean {
    return currentlyViewedFiles.has(filePath);
}

/**
 * Get all currently viewed files
 */
export function getCurrentlyViewedFiles(): Set<string> {
    return new Set(currentlyViewedFiles);
}

/**
 * Update the tree view to show which files are currently being viewed
 */
export function updateTreeViewIndicators(): void {
    const treeProvider = getCurrentTreeProvider();
    if (treeProvider) {
        // Refresh the tree to update the UI indicators
        treeProvider.refreshTreeIndicators();
    }
}

/**
 * Manually add a file to the currently viewed tracking
 * This is useful when programmatically opening files
 */
export function addCurrentlyViewedFile(filePath: string): void {
    currentlyViewedFiles.add(filePath);
    updateTreeViewIndicators();
}

/**
 * Manually remove a file from the currently viewed tracking
 */
export function removeCurrentlyViewedFile(filePath: string): void {
    currentlyViewedFiles.delete(filePath);
    updateTreeViewIndicators();
}

File: pycharm-extension/src/main/kotlin/com/github/haitham_almughrabi/rf_toolkit/logic/FileViewTracker.kt
package com.github.haitham_almughrabi.rf_toolkit.logic

import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.fileEditor.FileEditorManagerEvent
import com.intellij.openapi.fileEditor.FileEditorManagerListener
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import java.nio.file.Paths

/**
 * Tracks currently viewed files in the editor to show indicators in the import tree.
 * Handles multiple editors including split views.
 */
object FileViewTracker {
    
    private val currentlyViewedFiles = mutableSetOf<String>()
    private var updateCallback: (() -> Unit)? = null
    
    /**
     * Initialize the file view tracker for a project
     */
    fun initialize(project: Project, onUpdate: () -> Unit) {
        updateCallback = onUpdate
        
        // Initial update with currently open files
        updateCurrentlyViewedFiles(project)
        
        // Listen for editor changes
        project.messageBus.connect().subscribe(
            FileEditorManagerListener.FILE_EDITOR_MANAGER,
            object : FileEditorManagerListener {
                override fun fileOpened(source: FileEditorManager, file: VirtualFile) {
                    updateCurrentlyViewedFiles(project)
                }
                
                override fun fileClosed(source: FileEditorManager, file: VirtualFile) {
                    updateCurrentlyViewedFiles(project)
                }
                
                override fun selectionChanged(event: FileEditorManagerEvent) {
                    updateCurrentlyViewedFiles(event.manager.project)
                }
            }
        )
    }
    
    /**
     * Update the set of currently viewed files
     */
    private fun updateCurrentlyViewedFiles(project: Project) {
        currentlyViewedFiles.clear()
        
        val fileEditorManager = FileEditorManager.getInstance(project)
        
        // Get all open files (includes split views)
        val openFiles = fileEditorManager.openFiles
        for (file in openFiles) {
            // Normalize path for consistent tracking
            val normalizedPath = Paths.get(file.path).normalize().toString()
            currentlyViewedFiles.add(normalizedPath)
        }
        
        // Notify listeners that the view state has changed
        updateCallback?.invoke()
    }
    
    /**
     * Check if a file path is currently being viewed
     */
    fun isFileCurrentlyViewed(filePath: String): Boolean {
        val normalizedPath = Paths.get(filePath).normalize().toString()
        return currentlyViewedFiles.contains(normalizedPath)
    }
    
    /**
     * Get all currently viewed files
     */
    fun getCurrentlyViewedFiles(): Set<String> {
        return currentlyViewedFiles.toSet()
    }
    
    /**
     * Manually add a file to the currently viewed tracking
     */
    fun addCurrentlyViewedFile(filePath: String) {
        val normalizedPath = Paths.get(filePath).normalize().toString()
        currentlyViewedFiles.add(normalizedPath)
        updateCallback?.invoke()
    }
    
    /**
     * Manually remove a file from the currently viewed tracking
     */
    fun removeCurrentlyViewedFile(filePath: String) {
        val normalizedPath = Paths.get(filePath).normalize().toString()
        currentlyViewedFiles.remove(normalizedPath)
        updateCallback?.invoke()
    }
}
